sql/src/main/java/io/crate/planner/consumer/UpdatePlanner.java

```diff
@@ -196,8 +198,11 @@ private static ExecutionPlan updateByQuery(Functions functions,
         Reference idReference = requireNonNull(tableInfo.getReference(DocSysColumns.ID), "Table must have a _id column");
         Assignments assignments = Assignments.convert(assignmentByTargetCol);
         Symbol[] assignmentSources = assignments.bindSources(tableInfo, params, subQueryResults);
-        UpdateProjection updateProjection = new UpdateProjection(
-            new InputColumn(0, idReference.valueType()), assignments.targetNames(), assignmentSources, null);
+        UpdateProjection updateProjection = null;
+        if (tableInfo.isPartitioned() == false || tableInfo.partitions().isEmpty() == false) {
```

Comment:
Shouldn't the routing be empty in such a case - causing an "empty" collect operation anyway. Why did we receive any errors in such a case?

Reply:
I can add the explicit routing = EMPTY to be more clear, but the WHERE clause is a NO MATCH.
The error comes from https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java#L452 and https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java#L426

Reply:
What I don't get is why this routine that throws the error is even called. 

The fix here looks more like a workaround to me.

Reply:
Because we add a Delete/Update Projection to the Collect phase.

Reply:
Yep but these projections should have a `RowGranularity.SHARD`, which means that the projectors should only be created on a shard-level, which only should be possible if there are actual shards available.

Your explanation with the `NO_MATCH` for the "no concurrency" issue seems a bit odd as well. If that were the case this exception here in the planner wouldn't be necessary? Partitions could be deleted after the plan is created but before the plan is executed.